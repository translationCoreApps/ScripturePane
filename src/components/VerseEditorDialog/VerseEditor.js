import React from "react";
import PropTypes from 'prop-types';
import VerseEditorStepper from './VerseEditorStepper';
import EditScreen  from './screens/EditScreen';
import DoneIcon from 'material-ui/svg-icons/action/done';
import ReasonScreen from './screens/ReasonScreen';
import * as stepTypes from './steps';

const styles = {
  screen: {
    padding: '24px'
  },
  stepper: {
    borderBottom: 'solid 1px #999',
    height: '50px'
  },
  actions: {
    padding: '0 24px',
    textAlign: 'right'
  },
  icon: {
    color: '#ffffff',
    width: 20,
    height: 20,
    marginRight: 5,
    marginBottom: 5,
    verticalAlign: 'middle'
  }
};

const steps = ['Edit Verse', 'Select reason(s) for change'];

class VerseEditor extends React.Component {

  constructor(props) {
    super(props);
    this.handleBack = this.handleBack.bind(this);
    this.handleCancel = this.handleCancel.bind(this);
    this.handleNext = this.handleNext.bind(this);
    this.onLastStep = this.onLastStep.bind(this);
    this.state = {
      stepIndex: 0
    };
  }

  handleBack() {
    const {stepIndex} = this.state;
    this.setState({
      stepIndex: Math.max(stepIndex - 1, 0)
    });
  }

  handleCancel() {
    const {onCancel} = this.props;
    onCancel();
  }

  handleNext() {
    const {stepIndex} = this.state;
    const {onSubmit} = this.props;
    if(this.onLastStep()) {
      onSubmit();
    } else {
      this.setState({
        stepIndex: stepIndex + 1
      });
    }
  }

  onLastStep() {
    const {stepIndex} = this.state;
    return stepIndex === steps.length - 1;
  }

  render() {
    const {translate, onCancel} = this.props;
    const {stepIndex} = this.state;

    let screen;
    switch(stepIndex) {
      case 0:
        screen = (<ReasonScreen/>);
        break;
      case 1:
      default:
        screen = (<EditScreen/>);
    }

    let nextStepButtonTitle = translate('buttons.next_button');
    if(this.onLastStep()) {
      nextStepButtonTitle = (
        <React.Fragment>
          <DoneIcon style={styles.icon}/>
          {translate('buttons.save_button')}
        </React.Fragment>
      );
    }

    return (
      <div>
        <VerseEditorStepper stepIndex={stepIndex}
                            style={styles.stepper}
                            steps={steps}/>
        <div style={styles.screen}>
          {screen}
        </div>
        <div style={styles.actions}>
          <button className="btn"
                  disabled={stepIndex === 0}
                  onClick={this.handleBack}>
            {translate('buttons.back_button')}
          </button>
          <button className="btn-second"
                  onClick={onCancel}>
            {translate('buttons.cancel_button')}
          </button>
          <button className="btn-prime"
                  onClick={this.handleNext}>
            {nextStepButtonTitle}
          </button>
        </div>
      </div>
    );
  }
}

VerseEditor.propTypes = {
  translate: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

export default VerseEditor;
