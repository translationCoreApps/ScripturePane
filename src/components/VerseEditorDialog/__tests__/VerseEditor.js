/* eslint-env jest */

import React from 'react';
import VerseEditor, {isNextEnabled} from '../VerseEditor';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import {render} from 'enzyme';

describe('<VerseEditor/>', () => {


  it('renders', () => {
    const wrapper = render(
      <MuiThemeProvider>
        <VerseEditor translate={k=>k}
                     onCancel={jest.fn()}
                     onSubmit={jest.fn()}
                     verseText="verse text" />
      </MuiThemeProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });

});

describe('isNextEnabled', () => {
  it('is has not changed', () => {
    const state = {
      stepIndex: 0,
      verseChanged: false,
      newVerse: 'verse',
      reasons: ['reason']
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(false);
  });

  it('has an empty verse', () => {
    const state = {
      stepIndex: 0,
      verseChanged: true,
      newVerse: '',
      reasons: ['reason']
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(false);
  });

  it('has a null verse', () => {
    const state = {
      stepIndex: 0,
      verseChanged: true,
      newVerse: null,
      reasons: ['reason']
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(false);
  });

  it('has no reasons on step 1', () => {
    const state = {
      stepIndex: 0,
      verseChanged: true,
      newVerse: 'verse',
      reasons: []
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(true);
  });

  it('has no reasons on step 2', () => {
    const state = {
      stepIndex: 1,
      verseChanged: true,
      newVerse: 'verse',
      reasons: []
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(false);
  });

  it('has unknown step index', () => {
    const state = {
      stepIndex: 8,
      verseChanged: true,
      newVerse: 'verse',
      reasons: ['reason']
    };
    const result = isNextEnabled(state);
    expect(result).toEqual(false);
  });

});
