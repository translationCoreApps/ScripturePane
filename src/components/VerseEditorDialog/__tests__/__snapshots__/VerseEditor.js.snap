// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerseEditor/> renders 1`] = `
<div>
  <div
    style="display:flex;flex-direction:row;align-content:center;align-items:center;justify-content:space-between;border-bottom:solid 1px #999;height:50px;mui-prepared:"
  >
    <div
      style="flex:0 0 auto;mui-prepared:"
    >
      <span
        style="height:72px;color:rgba(0, 0, 0, 0.87);display:flex;align-items:center;font-family:Roboto, sans-serif;font-size:14px;padding-left:14px;padding-right:14px;font-weight:500;mui-prepared:"
      >
        <span
          style="padding-right:8px;mui-prepared:"
        >
          <svg
            style="display:block;color:#00bcd4;fill:#00bcd4;height:24px;width:24px;user-select:none;transition:all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms;font-size:24px;mui-prepared:"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <text
              fill="#fff"
              font-size="12"
              text-anchor="middle"
              x="12"
              y="16"
            >
              1
            </text>
          </svg>
        </span>
        Edit Verse
      </span>
    </div>
    <div
      style="flex:1 1 auto;mui-prepared:"
    >
      <span
        style="display:block;border-color:#bdbdbd;margin-left:-6px;border-top-style:solid;border-top-width:1px;mui-prepared:"
      />
    </div>
    <div
      style="flex:0 0 auto;margin-left:-6px;mui-prepared:"
    >
      <span
        disabled=""
        style="height:72px;color:rgba(0, 0, 0, 0.26);display:flex;align-items:center;font-family:Roboto, sans-serif;font-size:14px;padding-left:14px;padding-right:14px;cursor:default;mui-prepared:"
      >
        <span
          style="padding-right:8px;mui-prepared:"
        >
          <svg
            style="display:block;color:#9e9e9e;fill:#9e9e9e;height:24px;width:24px;user-select:none;transition:all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms;font-size:24px;mui-prepared:"
            viewBox="0 0 24 24"
          >
            <circle
              cx="12"
              cy="12"
              r="10"
            />
            <text
              fill="#fff"
              font-size="12"
              text-anchor="middle"
              x="12"
              y="16"
            >
              2
            </text>
          </svg>
        </span>
        Select reason(s) for change
      </span>
    </div>
  </div>
  <div
    style="padding:24px"
  >
    <textarea
      autofocus=""
      id="verse-editor-field"
      rows="4"
      style="width:100%;resize:none;padding:10px;border:solid 1px var(--border-color);font-style:inherit;font-variant:inherit;font-weight:inherit;font-stretch:inherit;font-size:inherit;line-height:inherit;font-family:inherit;cursor:inherit;outline:none;background-color:transparent;-webkit-appearance:textfield;color:rgba(0, 0, 0, 0.870588)"
    >
      verse text
    </textarea>
  </div>
  <div
    style="padding:0 24px;text-align:right"
  >
    <button
      class="btn"
      disabled=""
    >
      buttons.back_button
    </button>
    <button
      class="btn-second"
    >
      buttons.cancel_button
    </button>
    <button
      class="btn-prime"
      disabled=""
    >
      buttons.next_button
    </button>
  </div>
</div>
`;
